File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bfs/Problem16954.java
package com.keunwon.algorithm.bfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;

public class Problem16954 {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final int[][] moves = {{-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}, {-1, -1}, {0, 0}};

    private static char[][] board;

    public static void main(String[] args) throws IOException {
        input();
        bfs();
    }

    private static void bfs() {
        var queue = new LinkedList<Position>();
        queue.offer(new Position(7, 0));

        while (!queue.isEmpty()) {
            var size = queue.size();

            for (int i = 0; i < size; i++) {
                var p = queue.poll();

                if (board[p.x][p.y] == '#') { continue; }

                for (int[] move : moves) {
                    var nx = p.x + move[0];
                    var ny = p.y + move[1];

                    if (!validMap(nx, ny)) { continue; }

                    if (nx == 0 && ny == 7) {
                        System.out.println(1);
                        return;
                    }

                    if (board[nx][ny] != '#') {
                        queue.offer(new Position(nx, ny));
                    }
                }
            }

            down();
        }

        System.out.println(0);
    }

    private static void down() {
        for (int i = 7; i >= 1; i--) {
            System.arraycopy(board[i - 1], 0, board[i], 0, 8);
        }

        for (int i = 0; i < 8; i++) {
            board[0][i] = '.';
        }
    }

    private static boolean validMap(int x, int y) {
        return 0 <= x && x < 8 && 0 <= y && y < 8;
    }

    private static void input() throws IOException {
        board = new char[8][8];

        for (int i = 0; i < 8; i++) {
            var line = br.readLine();

            for (int j = 0; j < 8; j++) {
                board[i][j] = line.charAt(j);
            }
        }
    }

    private static class Position {
        int x;
        int y;

        Position(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem16954.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 움직이는 미로 탈출
 * Level: 골드-3
 **/
class Problem16954 {
    private lateinit var map: Array<CharArray>
    private val moves = arrayOf(-1 to -1, -1 to 0, -1 to 1, 0 to -1, 0 to 0, 0 to 1, 1 to -1, 1 to 0, 1 to 1)

    fun solution(map: Array<CharArray>): Int {
        this.map = map
        val queue = LinkedList<Node>().apply { offer(Node(7, 0)) }

        while (queue.isNotEmpty()) {
            val size = queue.size

            for (i in 0 until size) {
                val cur = queue.poll()

                if (map[cur.r][cur.c] == '#') continue

                for ((mr, mc) in moves) {
                    val nr = cur.r + mr
                    val nc = cur.c + mc

                    if (nr !in map.indices || nc !in map[0].indices) continue

                    if (nr == 0 && nc == 7) return 1

                    if (map[nr][nc] != '#') queue.offer(Node(nr, nc))
                }
            }
            down()
        }
        return 0
    }

    private fun down() {
        for (i in 6 downTo 0) {
            for (j in 0 until 8) {
                map[i + 1][j] = map[i][j]
            }
        }

        for (i in 0 until 8) {
            map[0][i] = '.'
        }
    }

    private data class Node(val r: Int, val c: Int)
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val map = Array(8) { br.readLine().toCharArray() }

        Problem16954().solution(map).also { bw.write("$it") }
        bw.flush()
    }
}
